kv.go

```diff
@@ -0,0 +1,436 @@
+}
+
+type kvs struct {
+	name   string
+	stream string
+	pre    string
+	js     *js
+}
+
+// Underlying entry.
+type kve struct {
+	bucket   string
+	key      string
+	value    []byte
+	revision uint64
+	delta    uint64
+	created  time.Time
+	op       KeyValueOp
+}
+
+func (e *kve) Bucket() string        { return e.bucket }
+func (e *kve) Key() string           { return e.key }
+func (e *kve) Value() []byte         { return e.value }
+func (e *kve) Revision() uint64      { return e.revision }
+func (e *kve) Created() time.Time    { return e.created }
+func (e *kve) Delta() uint64         { return e.delta }
+func (e *kve) Operation() KeyValueOp { return e.op }
+
+// Get returns the latest value for the key.
+func (kv *kvs) Get(key string) (KeyValueEntry, error) {
+	var b strings.Builder
+	b.WriteString(kv.pre)
+	b.WriteString(key)
+
+	m, err := kv.js.GetLastMsg(kv.stream, b.String())
+	if err != nil {
+		return nil, err
+	}
+
+	entry := &kve{
+		bucket:   kv.name,
+		key:      key,
+		value:    m.Data,
+		revision: m.Sequence,
+		created:  m.Time,
+	}
+
+	// Double check here that this is not a DEL Operation marker.
+	if len(m.Header) > 0 && m.Header.Get(kvop) == kvdel {
+		entry.op = KeyValueDelete
+		return entry, ErrMSgNotFound
+	}
+
+	return entry, nil
+}
+
+// Put will place the new value for the key into the store.
+func (kv *kvs) Put(key string, value []byte) (revision uint64, err error) {
+	var b strings.Builder
+	b.WriteString(kv.pre)
+	b.WriteString(key)
+
+	pa, err := kv.js.Publish(b.String(), value)
+	if err != nil {
+		return 0, err
+	}
+	return pa.Sequence, err
+}
+
+// Create will add the key/value pair iff it does not exist.
+func (kv *kvs) Create(key string, value []byte) (revision uint64, err error) {
+	v, err := kv.Update(key, value, 0)
+	if err == nil {
+		return v, nil
+	}
+	// TODO(dlc) - Since we have tombstones for DEL ops for watchers, this could be from that
+	// so we need to double check.
+	if e, err := kv.Get(key); err == ErrMSgNotFound {
+		return kv.Update(key, value, e.Revision())
+	}
+	return 0, err
+}
+
+// Update will update the value iff the latest revision matches.
+func (kv *kvs) Update(key string, value []byte, revision uint64) (uint64, error) {
+	var b strings.Builder
+	b.WriteString(kv.pre)
+	b.WriteString(key)
+
+	m := Msg{Subject: b.String(), Header: Header{}, Data: value}
+	m.Header.Set(ExpectedLastSubjSeqHdr, strconv.FormatUint(revision, 10))
+
+	pa, err := kv.js.PublishMsg(&m)
+	if err != nil {
+		return 0, err
+	}
+	return pa.Sequence, err
+}
+
+// Delete the key and all revisions.
+func (kv *kvs) Delete(key string) error {
+	var b strings.Builder
+	b.WriteString(kv.pre)
+	b.WriteString(key)
+
+	// DEL op marker. For watch functionality.
+	m := Msg{Subject: b.String(), Header: Header{}}
+	m.Header.Set(kvop, kvdel)
+	paf, err := kv.js.PublishMsgAsync(&m)
+	if err != nil {
+		return err
+	}
+	err = kv.js.purgeStream(kv.stream, &streamPurgeRequest{Subject: b.String(), Keep: 1})
```

Comment:
wouldnt using a future here and an unconditional purge end up in cases where the delete op message failed AND the stream got purged? that seems lke a bad outcome. The delete op is critical for watchers functioning and we can then follow that with a purge.  And we should purge up the sequence of the delete op, not to keep 1, since a write could have come in between the delete op and the purge.  Purging to the sequence of the delete op means no data will be lost in that case

Reply:
yes, just use a rollup

Reply:
Fixed server and updated.